### ISSUE TITLE
Add a step to the Issue New Card form that collects a magic code

### ISSUE THREAD

**Issue by tgolen** [Expensify/App#50667]
# Problem
Someone can issue a physical or virtual Expensify card without verifying they are the owner of the account. This relates to an [internal security issue](https://github.com/Expensify/Expensify/issues/434451).

## Why this is important to solve
This is a security vulnerability which can be taken advantage of if an account is compromised.

# Solution
Collect a magic code when issuing physical or virtual Expensify cards. In a little more detail:

- The two API commands are [here](https://github.com/Expensify/App/blob/0510a66ecf8d2003db3c9b295d90c7b10610d2c4/src/libs/actions/Card.ts#L636-L644)
- They should each have a new parameter called `validateCode` that is passed to the server
- In order to do this, the `IssueNewCardPage` needs a new step to gather a magic code from the user (we should have existing components that can be reused for this)

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Add a step to the Issue New Card form that collects a magic code

### What is the root cause of that problem?
This is a new feature request

### What changes do you think we should make in order to solve the problem?
- Add a validation step in `IssueNewCardPage` like `ValidationStep`. If this step is after the confirmation step, we should move the logic call API to the `ValidationStep`. If this step is before the confirmation step we will store the validation code data in `issueNewCard?.data` then pass it [here](https://github.com/Expensify/App/blob/cff3e862e3b7623d9ade431a64bac55645964a2a/src/pages/workspace/expensifyCard/issueNew/ConfirmationStep.tsx#L49)

https://github.com/Expensify/App/blob/cff3e862e3b7623d9ade431a64bac55645964a2a/src/pages/workspace/expensifyCard/issueNew/IssueNewCardPage.tsx#L33

- Add the `validationCode` param here

https://github.com/Expensify/App/blob/cff3e862e3b7623d9ade431a64bac55645964a2a/src/libs/actions/Card.ts#L628

- To create `ValidationStep`, we can use `ValidateCodeActionModal` with `isVisible` as `true`. The way to use it as we do [here](https://github.com/Expensify/App/blob/cff3e862e3b7623d9ade431a64bac55645964a2a/src/pages/settings/Wallet/ExpensifyCardPage.tsx#L314)

The details can be discussed in the PR phrase. I can create a test branch if we need
### What alternative solutions did you explore? (Optional)
NA

**Comment by tgolen** [comment]
@nkdengineer thanks! I was thinking the magic code should probably be the first step. My thinking is that there isn't much sense in letting the user fill out the rest of the form if they can't get past the magic code.

**Comment by nkdengineer** [comment]
@tgolen If we do that we need an API to help the user verrify before moving to the next step.

**Comment by nkdengineer** [comment]
- Using `ValidateCodeActionModal` is useful because 

1. It had the logic to call `requestValidateCodeAction` API when it was rendered
2. It had a separate file for the Android platform
3. It's also used in other places of the app

> If this step is after the confirmation step, we should move the logic call API to the ValidationStep

> To create ValidationStep, we can use ValidateCodeActionModal with isVisible as true. The way to use it as we do [here](https://github.com/Expensify/App/blob/cff3e862e3b7623d9ade431a64bac55645964a2a/src/pages/settings/Wallet/ExpensifyCardPage.tsx#L314)

- I notice that the idea of the last proposal is similar to what I proposed

**Comment by hungvu193** [comment]
> * Using `ValidateCodeActionModal` is useful because
> 
> 1. It had the logic to call `requestValidateCodeAction` API when it is rendered
> 2. It had a separate file for the Android platform
> 3. It's also used in other places of the app
> 
> > If this step is after the confirmation step, we should move the logic call API to the ValidationStep
> 
> > To create ValidationStep, we can use ValidateCodeActionModal with isVisible as true. The way to use it as we do [here](https://github.com/Expensify/App/blob/cff3e862e3b7623d9ade431a64bac55645964a2a/src/pages/settings/Wallet/ExpensifyCardPage.tsx#L314)
> 
> * I notice that the idea of the last proposal is similar to what I proposed
> * If we don't want to create a new step we can add `ValidateCodeActionModal` in the confirmation page and will open this modal when we click on the confirmation button

I also agree with @nkdengineer at this point. We also have a PR to reuse ValidateCodeModal as much as possible here. (https://github.com/Expensify/App/pull/49445)

**Comment by tgolen** [comment]
Yeah, that's right, we should reuse those existing components as much as possible. It sounds like this should work:

> If we don't want to create a new step we can add ValidateCodeActionModal in the confirmation page and will open this modal when we click on the confirmation button

**Comment by mountiny** [comment]
I want to highlight that @getusha and @hungvu193 are working on making the component for the magic code easier to reuse https://github.com/Expensify/App/pull/49445, but it has been proving tricky and taking a bit of time

**Comment by tgolen** [comment]
Thanks for the heads up @mountiny. Do you want to hold off on doing anything here until that's settled? Or is there something that we can do in this PR that maybe helps the progress of making it easier to reuse?

**Comment by mountiny** [comment]
I think it would make sense to wait on the refactor, or then we have to double the work. I hope the PR can be wrapped up soon. I have started a [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1729158252062769) for this here to hopefully get to faster completion.

**Comment by nkdengineer** [comment]
@tgolen as mention [here](https://github.com/Expensify/App/pull/50864#issuecomment-2431369466), the API still return successful response with invalid code so we need to change BE

**Comment by tgolen** [comment]
Oh, also, it looks like [the refactor PR](https://github.com/Expensify/App/pull/49445) is now merged, so you should be unblocked with your PR.

**Comment by tgolen** [comment]
@nkdengineer This should be ready for the frontend PR to be created.

**Comment by tgolen** [comment]
bump @nkdengineer are you able to get a PR going for this now?

**Comment by nkdengineer** [comment]
@tgolen I'm investigating some bugs in the [PR](https://github.com/Expensify/App/pull/50864#issuecomment-2452214182). Will give an update on Monday.

**Comment by tgolen** [comment]
@nkdengineer Do you have an update on this? If you're not able to prioritize this, I suggest you unassign it so that we can have someone else work on it.

**Comment by nkdengineer** [comment]
@tgolen I still update it daily, will give an update soon.

**Comment by hungvu193** [comment]
PR is still in progress

**Comment by tgolen** [comment]
PR was merged this morning.

**Comment by trjExpensify** [comment]
👋 Let's get the checklist done @hungvu193 ahead of Friday!

**Comment by hungvu193** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production
  - [ ] 2b. Reported on staging (deploy blocker)
  - [ ] 2c. Reported on both staging and production
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by trjExpensify** [comment]
Payment summary: 

- $250 to @hungvu193 (go ahead and request!)
- $250 to @nkdengineer (offer sent!)

**Comment by trjExpensify** [comment]
>IMO, this is a new feature, I don't think we need a regression test.

On this, I agree we don't need a new regression test [here](https://expensify.testrail.io/index.php?/cases/view/3272217&group_by=cases:section_id&group_order=asc&display_deleted_cases=0&group_id=314225). We have one for the issue card flow, and it would necessitate going through the magic code validation by design.

**Comment by trjExpensify** [comment]
Paid, closing!

**Comment by garrettmknight** [comment]
$250 approved for @hungvu193
